_posts/2022-01-21-dashboards-plugins-intro.markdown

```diff
@@ -0,0 +1,220 @@
+
+e.g. manifest file
+
+```json
+{
+  "id": "plugin_id",
+  "version": "1.0.0",
+  "opensearchDashboardsVersion": "opensearchDashboards",
+  "server": true,
+  "ui": true,
+  "requiredPlugins": ["data", "savedObjects"],
+  "optionalPlugins": []
+}
+```
+
+### Plugin definition
+
+All the remaining plugin logic is exposed via a single class that implements the `Plugin` interface. This is done so that it is easy for other plugins and the core system to discover and use a plugins features using a well defined API.
+
+#### Lifecycle methods
+
+Every plugin has 3 life cycle methods that the core system calls during the lifecycle of a plugin.
+
+- Setup: when the plugin is registered and initialized
+- Start: is where any "running" logic for your plugin would live. This only applies if you need to start listening for outside events (polling for work, listening on a port, etc.)
+- Stop: Used to cleanup runtime.
+
+e.g.
+
+```js
+// my_plugin/public/index.ts
+
+import { PluginInitializer } from '../../src/core/public';
+import { MyPlugin, MyPluginSetup, MyPluginStart } from './plugin';
+
+export const plugin: PluginInitializer<MyPluginSetup, MyPluginStart> = () => new MyPlugin();
+export {
+  MyPluginSetup,
+  MyPluginStart
+}
+
+
+// my_plugin/public/plugin.ts
+
+import { CoreSetup, CoreStart, Plugin } from '../../src/core/public';
+import { OtherPluginSetup, OtherPluginStart } from '../other_plugin';
+import { ThirdPluginSetup, ThirdPluginStart } from '../third_plugin';
+
+export interface MyPluginSetup {
+  registerThing(...);
+}
+
+export interface MyPluginStart {
+  getThing(): Thing;
+}
+
+export interface MyPluginSetupDeps {
+  otherPlugin: OtherPluginSetup;
+  thirdPlugin?: ThirdPluginSetup;  // Optional dependency
+}
+
+export interface MyPluginStartDeps {
+  otherPlugin: OtherPluginStart;
+  thirdPlugin?: ThirdPluginStart;  // Optional dependency
+}
+
+export class MyPlugin implements Plugin<
+  // All of these types are optional. If your plugin does not expose anything
+  // or depend on other plugins, these can be omitted.
+  MyPluginSetup,
+  MyPluginStart,
+  MyPluginSetupDeps,
+  MyPluginStartDeps,
+> {
+
+  public setup(core: CoreSetup, plugins: MyPluginSetupDeps) {
+    // should return a value that matches `MyPluginSetup`
+  }
+
+  public start(core: CoreStart, plugins: MyPluginStartDeps) {
+    // should return a value that matches `MyPluginStart`
+  }
+
+  public stop() { ... }
+}
+```
+
+## How does all this work?
+
+Now that I've talked about what makes up a plugin, let’s take a look at how it all works. The job of discovering, initializing and running plugins within OpenSearch Dashboards is handled by the [plugin service](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/core/server/plugins/plugins_service.ts) within the core system. It begins when we start OpenSearch Dashboards using `yarn start`. This kicks off the core workflow as follows:
```

Comment:
- *It begins when ~~we~~ **you**  start*
- `This kicks off the core workflow as follows:` I think you might want to change this a little bit. It's a bit unclear if this a developer workflow or what the core is doing. Maybe change to `The core performs the following actions`